Morphto v2 - with ajax support

pxpm opened the pull request:
## WHY

### BEFORE - What was wrong? What was happening before this PR?

Before everything ... PFEEEWWWW! 😫 done!

### AFTER - What is happening after this PR?

This is a re-implementation of #4330 but with support for ajax fields and a new syntax. 

There are examples of both usages (array and fluent) in the demo branch https://github.com/Laravel-Backpack/demo/pull/388 . fluent implemented, array documented. 

PRO pr is also updated: https://github.com/Laravel-Backpack/PRO/pull/18

@tabacitu can you give a round up on usability and let me know how we are here?

The only thing that I notice about usability is that the error messages are duplicated, in "morphTo" container field and the actual field:
![image](https://user-images.githubusercontent.com/7188159/182878915-325254b5-e0f8-41d6-b5e7-20547d740a6f.png)
Just need to discuss with you which one you think we should keep.

This was a wild wild ride! sooooooooo many details ... I LOVE it! 🙃 

**NOTE**: not implemented yet CRUD JS API.

tabacitu commented:
We decided to write docs for this in this PR, then try doing MorphToMany too, see how the docs would be different between the two.

pxpm commented:
Heads up @tabacitu this branch had been updated. 

MorphTo fields are now treated as subfields. 

![image](https://user-images.githubusercontent.com/7188159/183982493-4cb049cb-0f24-4eac-a745-fd8e802704d1.png)

I've updated demo and pro PR too. Also added to docs: https://github.com/Laravel-Backpack/docs/pull/368

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -358,4 +408,132 @@ private function modelMethodIsRelationship($model, $method)
+        return [
+            'name' => $morphTypeFieldName,
+            'type' => 'relationship.morphTo_type_select',
+            'placeholder' => 'Select the '.$relationName,
```
I don't think it makes sense for the admin to write out the relation name. It will look super-weird to him. Yes, the developer might know what a `commentable` is... but the admin does not:

![CleanShot 2022-08-29 at 19 14 38](https://user-images.githubusercontent.com/1032474/187246676-c0e7f131-cb60-42b7-b805-c7b1edfab53c.png)


The developer (if they're any good) should hide that abstractization from the admin. I think the solution here is simple:

```suggestion
            'placeholder' => 'Select an entry',
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -358,4 +408,132 @@ private function modelMethodIsRelationship($model, $method)
+            'name' => $morphIdFieldName,
+            'type' => 'relationship.morphTo_select',
+            'entity' => false,
+            'placeholder' => 'Select the '.$relationName,
```
```suggestion
            'placeholder' => 'Select an entry',
```

jorgetwgroup commented:
Hi @pxpm i dont know if i do something wrong. Here we go:

I set my branches as:

PRO: morphTo-relation
CRUD: morphto-v2
DEMO: morphto

But when i go to CRUD: admin/pet-shop/comment/create i get 500 error

![Screen Shot 2022-09-12 at 09 29 18](https://user-images.githubusercontent.com/44643608/189655476-5ba6d994-459e-4fa0-8fa7-1e7903f7bdbf.png)

I get error details and is about memory leak:

2022/09/12 12:15:24 [error] 10#10: *38 FastCGI sent in stderr: "PHP message: PHP Deprecated:  Required parameter $options follows optional parameter $label in /var/www/bp_demo_20220803/vendor/backpack/crud/src/app/Library/CrudPanel/Traits/Relationships.php on line 429PHP message: PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /var/www/bp_demo_20220803/vendor/laravel/framework/src/Illuminate/Database/Connection.php on line 420PHP message: PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 65536 bytes) in /var/www/bp_demo_20220803/vendor/composer/ClassLoader.php on line 571" while reading response header from upstream, client: 172.19.0.1, server: backpack.test, request: "GET /admin/pet-shop/comment/create HTTP/1.1", upstream: "fastcgi://172.19.0.5:9000", host: "backpack.test", referrer: "http://backpack.test/admin/pet-shop/comment"


2022/09/12 12:20:21 [error] 10#10: *40 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 20480 bytes) in /var/www/bp_demo_20220803/vendor/laravel/framework/src/Illuminate/Database/Connection.php on line 420PHP message: PHP Fatal error:  Allowed memory size of 268435456 bytes exhausted (tried to allocate 65536 bytes) in /var/www/bp_demo_20220803/vendor/composer/ClassLoader.php on line 571" while reading response header from upstream, client: 172.19.0.1, server: backpack.test, request: "GET /admin/pet-shop/comment/create HTTP/1.1", upstream: "fastcgi://172.19.0.5:9000", host: "backpack.test", referrer: "http://backpack.test/admin/pet-shop/comment"

I have not limit set on PHP.

Cheers.

jorgetwgroup commented:
Hi @pxpm problem was created because monster have more than 500K rows, now is working.

I check implementation, till now look working normal, but i have a commment about error message: "The commentable.commentable id field is required."

![Screen Shot 2022-09-12 at 15 14 42](https://user-images.githubusercontent.com/44643608/189727010-1c19d80d-7d84-43dd-bc62-57da9b859797.png)

To prevent someone need to be force to translate, is possible a message like "The commentable in commentable id field is required." replace "." by "in".

Let me know if this is possible or what you think about it.

ty.

Cheers.

pxpm commented:
I've change the error message, now it reads: `This field is requiried`. 
I think writting `commentable` and technical terms here does not help UI. 

This is a message for a non-developer user. 

Cheers

jorgetwgroup commented:
Hello @pxpm im back :)

Look great for me, now is not needed touch request.

![screenshot-backpack test-2022 09 22-16_40_28](https://user-images.githubusercontent.com/44643608/191836467-c4d1a9e6-312d-4f8d-8c90-e8d763e839cd.png)

Cheers.

tabacitu commented on src/app/Library/CrudPanel/Traits/Fields.php:
```diff
@@ -56,6 +56,18 @@ public function makeSureFieldHasNecessaryAttributes($field)
 
         $this->setupFieldValidation($field, $field['parentFieldName'] ?? false);
 
+        if (isset($field['relation_type']) && $field['relation_type'] === 'MorphTo') {
+            [$morphTypeFieldName, $morphIdFieldName] = $this->getMorphToFieldNames($field['name']);
+            if (! $this->hasFieldWhere('name', $morphTypeFieldName) || ! $this->hasFieldWhere('name', $morphIdFieldName)) {
+                // create the morph fields in the crud panel
+                $field = $this->createMorphToRelationFields($field, $morphTypeFieldName, $morphIdFieldName);
+                foreach ($field['morphOptions'] ?? [] as $morphOption) {
+                    [$key, $label, $options] = $this->getMorphOptionStructured($morphOption);
+                    $field = $this->addMorphOption($field, $key, $label, $options);
+                }
+            }
+        }
```
Ugh... this kind of pollutes this function, doesn't it? Shouldn't we move all this logic to a private function, something like `makeSureMorphSubfieldsAreDefined($field)`?

tabacitu replied:
Additionally, why do we do this AFTER we call `setupFieldValidation()` and not BEFORE it? Any reason?

pxpm replied:
No special reason that I can recall, it was just so that this block of code was not messing with the clean defined stuff. 

Moving this to a private method solved that issue. 

Cheers

tabacitu commented on src/app/Library/CrudPanel/Traits/Input.php:
```diff
@@ -49,6 +49,7 @@ private function getDirectInputsFromInput($input, $model = false, $fields = [],
 
         $input = $this->decodeJsonCastedAttributes($input, $model);
         $input = $this->compactFakeFields($input, $model, $fields);
+        $input = $this->addMorphToInputsFromRelationship($input);
```
Now this is nice and clean.

tabacitu replied:
Perhaps a better name for this would be "include" if the next method is "exclude"?

```suggestion
        $input = $this->includeMorphToInputsFromRelationship($input);
```

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -262,6 +283,35 @@ public function getRelationFieldsWithPivot()
+     * @param  string  $relationName  the morphto relation name
+     * @return array
+     */
+    public function getMorphToFieldNames(string $relationName)
```
Any reason for this to be `public`? I only see it used within CrudPanel traits.

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -358,4 +408,132 @@ private function modelMethodIsRelationship($model, $method)
+     * @param  array  $options
+     * @return void|array
+     */
+    public function addMorphOption($fieldName, string $key, $label = null, array $options = [])
```
Would work just as well as `protected` but disallow editing, correct?

```suggestion
    protected function addMorphOption($fieldName, string $key, $label = null, array $options = [])
```

pxpm replied:
This can't be protected or private otherwise `CrudField` class couldn't call it. (CrudField is an "outsider" here, from our conversation on discord the other day).
ping @tabacitu

tabacitu replied:
Damn it! Ok...

tabacitu commented on src/app/Library/CrudPanel/Traits/Relationships.php:
```diff
@@ -192,6 +192,27 @@ private function changeBelongsToNamesFromRelationshipToForeignKey($input, $field
         return $input;
     }
 
+    /**
```
Maybe it'd make sense to move all these new Morph functions to a new CrudPanel trait... say `MorphRelationships`? 🤷‍♂️ What do you think @pxpm ?

tabacitu commented on src/app/Library/CrudPanel/Traits/MorphRelationships.php:
```diff
@@ -0,0 +1,210 @@
+     * @param  array  $fields
+     * @return array
+     */
+    private function addMorphToInputsFromRelationship($input)
```
```suggestion
    private function includeMorphToInputsFromRelationship($input)
```

tabacitu commented on src/app/Library/CrudPanel/Traits/MorphRelationships.php:
```diff
@@ -0,0 +1,210 @@
+    /**
+     * This function is responsible for setting up the morph fields structure.
+     * Developer can define the morph structure as follows:
+     *  'morphOptions => [
+     *       ['nameOnAMorphMap', 'label', [options]],
+     *       ['App\Models\Model'], // display the name of the model
+     *       ['App\Models\Model', 'label', ['data_source' => backpack_url('smt')]
+     *  ]
+     * OR
+     * ->addMorphOption('App\Models\Model', 'label', ['data_source' => backpack_url('smt')]).
```
I like this syntax explained in the comments, but it's confusing to have it here. That's not the syntax for THIS method, it's the syntax for the same method on the CrudField object. I only tracked that down after wanting to point out this method's signature has one more parameter 😀

So let's move this comment there please. And write out for each parameter what it actually is.

tabacitu commented on src/app/Library/CrudPanel/Traits/MorphRelationships.php:
```diff
@@ -0,0 +1,210 @@
+                $key = $morphMap[array_search($key, $morphMap)];
+
+                if (array_key_exists($key, $morphTypeField['options'])) {
+                    throw new \Exception('Duplicate entry for «'.$key.'» in morphOptions');
```
This error message is the same 3 times. That should never happen, it defeats the purpose of error messages. If someone comes back to us saying "_i got this error message_" we won't be able to tell exactly where it errored, right? Let's make these error messages more specific, so that they pinpoint what went wrong, or recommend specifically what to fix. The way I see it, each error message should be slightly different, otherwise... it would have been caught by the previous exception, right?

tabacitu commented on src/app/Library/CrudPanel/Traits/MorphRelationships.php:
```diff
@@ -0,0 +1,210 @@
+     * @param  array  $options
+     * @return void|array
+     */
+    public function addMorphOption($fieldName, string $key, $label = null, array $options = [])
```
This method is 50 lines long, and does a lot of stuff. So let's please add comments before each thing we do.

pxpm commented:
@tabacitu did the changes. 

Just one question still un-answered https://github.com/Laravel-Backpack/CRUD/pull/4579#discussion_r992215203

Cheers

tabacitu commented:
<img width="934" alt="CleanShot 2022-10-14 at 16 43 13@2x" src="https://user-images.githubusercontent.com/1032474/195861761-da196026-1cd7-4dc5-a624-6982c6e66fc0.png">

OMG PEDRO IT'S HAPPENING!!! 😱

tabacitu commented:
Here's the GIF for it:

![CleanShot 2022-11-23 at 17 02 09](https://user-images.githubusercontent.com/1032474/203553594-4be3523b-c93a-4bc1-b6ee-a4184383bc5a.gif)

pxpm commented:
> Here's the GIF for it:
> 
> ![CleanShot 2022-11-23 at 17 02 09](https://user-images.githubusercontent.com/1032474/203553594-4be3523b-c93a-4bc1-b6ee-a4184383bc5a.gif) [ ![CleanShot 2022-11-23 at 17 02 09](https://user-images.githubusercontent.com/1032474/203553594-4be3523b-c93a-4bc1-b6ee-a4184383bc5a.gif) ](https://user-images.githubusercontent.com/1032474/203553594-4be3523b-c93a-4bc1-b6ee-a4184383bc5a.gif) [ ](https://user-images.githubusercontent.com/1032474/203553594-4be3523b-c93a-4bc1-b6ee-a4184383bc5a.gif)

A gif with a "broken/failing" test doesn't seem very appealing to me @tabacitu 💥
